### ISSUE TITLE
[pooling] The kernels hit system limits when window size is large

### ISSUE THREAD

**Issue by atamazov** [ROCmSoftwarePlatform/MIOpen#2110]
MIOpen runs into build errors, the issue is reported at https://github.com/microsoft/onnxruntime/issues/15482. The issue is not FP16-specific.

### (1) Forward
```bash
$ ./bin/MIOpenDriver poolfp16 -M 0 --input 1x64x410x400,10496000x164000x400x1 -y 410 -x 400 -p 0 -q 0 -v 1 -u 1 -m avg -F 1 -t 1 -i 1
...
error: stack frame size (328004) exceeds limit (262112) in function 'mloPoolingG'
```
### (2) Backward
```bash
$ ./bin/MIOpenDriver poolfp16 -M 0 --input 1x3x255x255,195075x65025x255x1 -y 255 -x 255 -p 0 -q 0 -v 1 -u 1 -m avg -F 0 -t 1 -i 1
...
error: local memory (137288) exceeds limit (65536) in function 'mloPoolingAveBwd'
error: local memory (205940) exceeds limit (65536) in function 'mloPoolingMaxBwd'
```
## Root reason

The reason is too big sizes of private local arrays in kernels, e.g. those ones:
https://github.com/ROCmSoftwarePlatform/MIOpen/blob/c5049aa71f910c2d798b8c40ea21786669c41625/src/kernels/MIOpenPooling.cl#L83-L84

----
[Attribution] @junliume @johnny-keker
- https://github.com/ROCmSoftwarePlatform/MIOpen/labels/enhancement
- https://github.com/ROCmSoftwarePlatform/MIOpen/labels/urgency_blocker
- https://github.com/ROCmSoftwarePlatform/MIOpen/labels/value_high
- Proposed assignees:
  - @JehandadKhan
  - @ce1adon 
  - @carlushuang
  - @asroy
  - @atamazov

**Comment by atamazov** [comment]
## How to fix (proposal)

- (1) At first, provide an alternative "naive" solver(s)/kernel(s). It should back up cases when we hit stack frame of local memory limit. The open question (except the naive kernel itself) is how to reliably find out what are the limits (maybe these depend on the target HW).
  - - [x] AP (1.1) Forward
      - #2163
  - - [x] AP (1.2) Backward - all except workspace index mask mode for Max Pooling
      - #2372
  - - [ ] AP (1.3) Resolve [#2371](https://github.com/ROCmSoftwarePlatform/MIOpen/issues/2371). This is necessary to support workspace index mask in the solver that supports wide 2D pooling window.
  - Note that this solution would not lead to performance regression, because pooling does not work at all right now, but the performance will be worse than GPU may potentially deliver.

- (2) The reasonable second step is designing and implementing the new pooling solver(s) that should have high implementation limits but deliver better performance than naive implementation. This is optimization task which is out of the scope of this ticket.

**Comment by dmikushin** [comment]
@atamazov , These are local arrays in terms of HIP, not in terms of OpenCL, right? So they are automatic arrays, private to each thread. In this case, I'm wondering how it could be limited. Theoretically, local memory is infinite, but large sizes will quickly degrade the performance. I remember LLVM/Clang emitting "LOCAL_DEPOT" - a special array that aggregates all emitted local data of a GPU kernel. Perhaps, compiler sets out a limit on its size, which could be lifted?

**Comment by atamazov** [comment]
AP 1.2 (+more) is resolved by #2372
